File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.py
class Solution:
    def smallestPalindrome(self, s: str) -> str:
        n = len(s)
        freq = [0] * 26

        # Count frequencies in the first half
        for i in range(n // 2):
            freq[ord(s[i]) - 97] += 1

        arr = list(s)
        j = 0
        for i in range(26):
            while freq[i]:
                arr[j] = arr[n - 1 - j] = chr(97 + i)
                j += 1
                freq[i] -= 1

        return ''.join(arr)

File: 3517-smallest-palindromic-rearrangement-i/3517-smallest-palindromic-rearrangement-i.cpp
class Solution {
public:
    string smallestPalindrome(string s) {
        int n = s.size(), j = 0;
        int freq[26]{};

        // Count frequencies of characters in the first half
        for (int i = 0; i < n >> 1; i++)            
            freq[(s[i] & 31) - 1]++;      // ASCII → char index

        // Fill the string symmetrically with sorted characters
        for (int i = 0; i < 26; i++) {
            while (freq[i]--) {                
                s[j] = 'a' + i;           // left side
                s[n - 1 - j++] = 'a' + i; // right side
            }
        }

        return s;
    }
};